clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
Ah, I think I was wrong on one point. Your initial patch discarded ALL data until the exception was thrown. That would resolve the issue I mentioned above. That approach is easier to reason about if we could find a way to collapse the loops to remove some of its ugliness. We might also want to skip sending new fetches until the exception was handled to avoid fetching data we know will be discarded anyway. That would have implications for pollOnce (similar as above). Haha, not sure either approach is safe from complexity.